### PR TITLE
[msal-node] Cache-3: Add NodeStorage, token generation and separate SPAresponse

### PR DESCRIPTION
**Pull request by sameerag**
This PR is third in the series of cache PRs. Since I am breaking down a bigger PR into consumable chunks, this may not work end to end but does the following changes:
* Defines `NodeStorage` with two new base APIs to read and write to external cache
* A temporary `NodeCacheManager` created on request from AzureSDK which will soon morph into `keytar` hook and eventually the extension library
* A `UnifiedCacheManager`, currently does getAccounts and adds credentials to cache. This file will expand in future for cache lookups.
* `TokenGenerator` to generate the token entities from response data
* Separated `SPAResponseHandler` for now ; this will fold to a single file in `msal-common` soon

Tests:
Tests are coming, they are 80% there, need a few more

### CODE REVIEW

File: lib/msal-node/src/cache/NodeCacheManager.ts [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+export class NodeCacheManager {
+    cache: string;
+
+    /**
+     * Read contents of the cache blob to in memoryCache
+     * @param cachePath
+     */
+    readFromFile(cacheLocation: string): string {
+        this.cache = fs.readFileSync(cacheLocation, 'utf8');
```
Comment (jasonnutter) [inline lib/msal-node/src/cache/NodeCacheManager.ts:16]: Can you make this async? https://nodejs.org/api/fs.html#fs_fs_promises_api

File: lib/msal-node/src/cache/NodeCacheManager.ts [hunk 2]
```diff
@@ -0,0 +1,27 @@
+/*
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+const fs = require('fs');
```
Comment (jasonnutter) [inline lib/msal-node/src/cache/NodeCacheManager.ts:6]: We should use `import` instead of require everywhere by default.

File: lib/msal-node/src/cache/NodeCacheManager.ts [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+export class NodeCacheManager {
+    cache: string;
```
Comment (jasonnutter) [inline lib/msal-node/src/cache/NodeCacheManager.ts:9]: Why is this needed?

File: lib/msal-node/src/cache/NodeCacheManager.ts [hunk 4]
```diff
@@ -0,0 +1,27 @@
+/*
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+const fs = require('fs');
+
+export class NodeCacheManager {
```
Comment (jasonnutter) [inline lib/msal-node/src/cache/NodeCacheManager.ts:8]: I think just `CacheManager` is fine, don't need `Node`.

File: lib/msal-node/src/cache/NodeStorage.ts [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+     * retrieve the file Path to read the cache from
+     */
+    getCachePath(): string {
+        return this.cachePath;
+    }
+
+    /**
+     * read JSON formatted cache from disk
+     */
+    getSerializedCache(): string {
```
Comment (jasonnutter) [inline lib/msal-node/src/cache/NodeStorage.ts:37]: I think these should be async, since file IO should be async.

File: lib/msal-common/src/response/SPAResponseHandler.ts [hunk 6]
```diff
@@ -26,7 +26,7 @@ import { ServerError } from "../error/ServerError";
 /**
  * Class that handles response parsing.
  */
-export class ResponseHandler {
+export class SPAResponseHandler {
```
Comment (jasonnutter) [inline lib/msal-common/src/response/SPAResponseHandler.ts:29]: These changes seem unrelated to the `NodeStorage` changes, could this be split into two PRs?

File: lib/msal-common/src/unifiedCache/UnifiedCacheManager.ts [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+
+        const accountKey = accountCacheKey
+            .join(Separators.CACHE_KEY_SEPARATOR)
+            .toLowerCase();
+
+        if (this.inMemoryCache.accounts[accountKey])
+            return this.inMemoryCache.accounts[accountKey];
+        else {
+            return null;
+        }
```
Comment (jasonnutter) [inline lib/msal-common/src/unifiedCache/UnifiedCacheManager.ts:92]:
```js
return this.inMemoryCache.accounts[accountKey] || null
```

File: lib/msal-common/src/unifiedCache/UnifiedCacheManager.ts [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+    /**
+     * append credential cache to in memory cache
+     * @param idT: IdTokenEntity
+     * @param at: AccessTokenEntity
+     * @param rt: RefreshTokenEntity
+     */
+    addCredentialCache(
+        at: AccessTokenEntity,
+        idT: IdTokenEntity,
+        rt: RefreshTokenEntity
```
Comment (jasonnutter) [inline lib/msal-common/src/unifiedCache/UnifiedCacheManager.ts:104]: Would prefer `accessToken`, `idToken`, and `refreshToken` instead of the short names.

File: lib/msal-common/src/client/SPAClient.ts [hunk 9]
```diff
@@ -415,12 +415,12 @@ export class SPAClient extends BaseClient {
...
 
         // Create response handler
-        const responseHandler = new ResponseHandler(this.clientConfig.auth.clientId, this.cacheStorage, this.cacheManager, this.cryptoUtils, this.logger);
+        const responseHandler = new SPAResponseHandler(this.clientConfig.auth.clientId, this.cacheStorage, this.cacheManager, this.cryptoUtils, this.logger);
         // Validate response. This function throws a server error if an error is returned by the server.
         responseHandler.validateServerAuthorizationTokenResponse(acquiredTokenResponse.body);
         // Return token response with given parameters
         const tokenResponse = responseHandler.createTokenResponse(acquiredTokenResponse.body, tokenRequest.authority, tokenRequest.resource, codeResponse && codeResponse.userRequestState);
-        // Set current account to received response account, if any.
+        // Set current account to received response accoßunt, if any.
```
Comment (sangonzal) [inline lib/msal-common/src/client/SPAClient.ts:423]: nit: weird char in account comment.

File: lib/msal-common/src/unifiedCache/UnifiedCacheManager.ts [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+
+    constructor(cacheImpl: ICacheStorage) {
+        this.cacheStorage = cacheImpl;
+        this.inMemoryCache = this.generateInMemoryCache(
+            this.cacheStorage.getSerializedCache()
+        );
+    }
+
+    /**
+     * set the cache memory
```
Comment (sangonzal) [inline lib/msal-common/src/unifiedCache/UnifiedCacheManager.ts:30]: nit: Consider only leaving comments where they add extra information. For example this comment is exactly the same thing as the setter name - it adds not extra value and just clutters the code.

File: lib/msal-common/src/unifiedCache/UnifiedCacheManager.ts [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+     */
+    generateInMemoryCache(cache: string): InMemoryCache {
+        return Deserializer.deserializeAllCache(
+            Deserializer.deserializeJSONBlob(cache)
+        );
+    }
+
+    /**
+     * retrieves the final JSON
+     * TODO: move this to msal-common
```
Comment (sangonzal) [inline lib/msal-common/src/unifiedCache/UnifiedCacheManager.ts:54]: nit: seems TODO is outdated

File: lib/msal-common/src/unifiedCache/generator/TokenCacheGenerator.ts [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+     */
+    static createIdTokenEntity(
+        homeAccountId: string,
+        authenticationResult: AuthenticationResult,
+        clientId: string,
+        environment: string
+    ): IdTokenEntity {
+        const idTokenEntity = new IdTokenEntity();
+
+        idTokenEntity.credentialType = "IdToken";
```
Comment (sangonzal) [inline lib/msal-common/src/unifiedCache/generator/TokenCacheGenerator.ts:27]: Maybe credentialType should be an Enum?

File: lib/msal-node/src/cache/NodeStorage.ts [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { ICacheStorage } from '@azure/msal-common';
+import { NodeCacheManager } from './NodeCacheManager';
+import { CacheOptions } from '../config/ClientConfiguration';
+
+/**
+ * This class implements the cache storage interface for MSAL through browser local or session storage.
```
Comment (sangonzal) [inline lib/msal-node/src/cache/NodeStorage.ts:11]: This comment doesn't apply here.

File: lib/msal-common/src/unifiedCache/generator/TokenCacheGenerator.ts [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+
+export class TokenCacheGenerator {
+    /**
+     * Create IdTokenEntity
+     * @param homeAccountId
+     * @param authenticationResult
+     * @param clientId
+     * @param authority
+     */
+    static createIdTokenEntity(
```
Comment (sangonzal) [inline lib/msal-common/src/unifiedCache/generator/TokenCacheGenerator.ts:19]: Maybe these static methods should be part of the entity classes themselves?

File: lib/msal-common/src/unifiedCache/generator/TokenCacheGenerator.ts [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { AccessTokenEntity } from "../entities/AccessTokenEntity";
+import { IdTokenEntity } from "../entities/IdTokenEntity";
+import { RefreshTokenEntity } from "../entities/RefreshTokenEntity";
+import { AuthenticationResult } from "../../response/AuthenticationResult";
+
+export class TokenCacheGenerator {
```
Comment (sangonzal) [inline lib/msal-common/src/unifiedCache/generator/TokenCacheGenerator.ts:11]: nit: TokenCacheEntityFactory

File: lib/msal-node/src/cache/NodeCacheManager.ts (hunk 4 above)
Comment (sameerag) [reply lib/msal-node/src/cache/NodeCacheManager.ts:8]: Done

File: lib/msal-common/src/unifiedCache/UnifiedCacheManager.ts (hunk 8 above)
Comment (sameerag) [reply lib/msal-common/src/unifiedCache/UnifiedCacheManager.ts:104]: Done

File: lib/msal-common/src/client/SPAClient.ts (hunk 9 above)
Comment (sameerag) [reply lib/msal-common/src/client/SPAClient.ts:410]: edited. Thanks.

File: lib/msal-common/src/unifiedCache/UnifiedCacheManager.ts (hunk 10 above)
Comment (sameerag) [reply lib/msal-common/src/unifiedCache/UnifiedCacheManager.ts:30]: This is mostly function description, to follow the norms describing each function. Choosing to comment a few functions and not others looks odd to me. We can clean up unneeded comments within code but I would still like each function have a little desc.

File: lib/msal-common/src/unifiedCache/UnifiedCacheManager.ts (hunk 11 above)
Comment (sameerag) [reply lib/msal-common/src/unifiedCache/UnifiedCacheManager.ts:54]: Removed. Thanks.

File: lib/msal-common/src/unifiedCache/generator/TokenCacheGenerator.ts (hunk 14 above)
Comment (sameerag) [reply lib/msal-common/src/unifiedCache/generator/TokenCacheGenerator.ts:19]: Yes. This is a good idea. Moving them.

File: lib/msal-node/src/cache/NodeStorage.ts (hunk 13 above)
Comment (sameerag) [reply lib/msal-node/src/cache/NodeStorage.ts:11]: Corrected.

File: lib/msal-common/src/unifiedCache/generator/TokenCacheGenerator.ts (hunk 15 above)
Comment (sameerag) [reply lib/msal-common/src/unifiedCache/generator/TokenCacheGenerator.ts:11]: This does not exist anymore as all token generation logic is pushed to the corresponding entity files

File: lib/msal-common/src/unifiedCache/UnifiedCacheManager.ts (hunk 7 above)
Comment (sameerag) [reply lib/msal-common/src/unifiedCache/UnifiedCacheManager.ts:92]: Done

File: lib/msal-common/src/response/SPAResponseHandler.ts (hunk 6 above)
Comment (sameerag) [reply lib/msal-common/src/response/SPAResponseHandler.ts:30]: Yes. I am retaining this as the PRs following up are dependent on this change, but will adhere to this norm in creating future PRs.

File: lib/msal-node/src/cache/NodeCacheManager.ts (hunk 3 above)
Comment (sameerag) [reply lib/msal-node/src/cache/NodeCacheManager.ts:9]: Yep, you are right. Not needed.

File: lib/msal-node/src/cache/NodeCacheManager.ts (hunk 2 above)
Comment (sameerag) [reply lib/msal-node/src/cache/NodeCacheManager.ts:6]: Done. Using `fspromises` now.

File: lib/msal-node/src/cache/NodeCacheManager.ts (hunk 1 above)
Comment (sameerag) [reply lib/msal-node/src/cache/NodeCacheManager.ts:16]: Done

File: lib/msal-node/src/cache/NodeStorage.ts (hunk 5 above)
Comment (sameerag) [reply lib/msal-node/src/cache/NodeStorage.ts:37]: Done.

File: lib/msal-common/src/unifiedCache/entities/AccessTokenEntity.ts [hunk 16]
```diff
@@ -34,4 +35,45 @@ export class AccessTokenEntity extends Credential {
...
+        atEntity.secret = authenticationResult.accessToken;
+
+        const date = new Date();
+        const currentTime = date.getMilliseconds() / 1000;
+        atEntity.cachedAt = currentTime.toString();
+
+        // TODO: Crosscheck the exact conversion UTC
+        // Token expiry time.
+        // This value should be  calculated based on the current UTC time measured locally and the value  expires_in Represented as a string in JSON.
+        atEntity.expiresOn = authenticationResult.expiresOn
```
Comment (pkanher617) [inline lib/msal-common/src/unifiedCache/entities/AccessTokenEntity.ts:65]: Remember that in the past this was returned as seconds, not milliseconds.

File: lib/msal-common/src/unifiedCache/entities/AccessTokenEntity.ts [hunk 17]
```diff
@@ -34,4 +35,45 @@ export class AccessTokenEntity extends Credential {
...
+            .getMilliseconds()
+            .toString();
+        atEntity.extendedExpiresOn = authenticationResult.extExpiresOn
```
Comment (pkanher617) [inline lib/msal-common/src/unifiedCache/entities/AccessTokenEntity.ts:68]: Same comment as above about seconds vs milliseconds

File: lib/msal-common/src/unifiedCache/entities/IdTokenEntity.ts [hunk 18]
```diff
@@ -27,4 +28,29 @@ export class IdTokenEntity extends Credential {
...
     }
+
+    /**
+     * Create IdTokenEntity
+     * @param homeAccountId
+     * @param authenticationResult
+     * @param clientId
+     * @param authority
+     */
+    static createIdTokenEntity(
```
Comment (pkanher617) [inline lib/msal-common/src/unifiedCache/entities/IdTokenEntity.ts:39]: Is this class a replacement for the IdToken and IdTokenClaims class, or is it just used by the cache?

File: lib/msal-node/src/cache/Storage.ts [hunk 19]
```diff
@@ -2,230 +2,101 @@
...
+import { CacheManager } from './CacheManager';
+import { CacheOptions } from '../config/ClientConfiguration';
 
 /**
- * This class implements the cache storage interface for MSAL through browser local or session storage.
- * Cookies are only used if storeAuthStateInCookie is true, and are only used for
- * parameters such as state and nonce, generally.
+ * This class implements Storage for node, reading cache from user specified storage location or an  extension library
  */
 export class Storage implements ICacheStorage {
```
Comment (pkanher617) [inline lib/msal-node/src/cache/Storage.ts:13]: Should this be named NodeStorage?

File: lib/msal-common/src/unifiedCache/entities/IdTokenEntity.ts (hunk 18 above)
Comment (sameerag) [reply lib/msal-common/src/unifiedCache/entities/IdTokenEntity.ts:39]: Just used by the cache. `IdToken` and `IdTokenClaims` classes will remain.

File: lib/msal-common/src/unifiedCache/entities/AccessTokenEntity.ts (hunk 16 above)
Comment (sameerag) [reply lib/msal-common/src/unifiedCache/entities/AccessTokenEntity.ts:65]: yes. Will correct in silent flow logic.

File: lib/msal-node/src/cache/Storage.ts (hunk 19 above)
Comment (sameerag) [reply lib/msal-node/src/cache/Storage.ts:12]: apparently "storage" is more intuitive as per other feedback. We can revisit this later.

File: lib/msal-common/src/unifiedCache/generator/TokenCacheGenerator.ts (hunk 12 above)
Comment (sameerag) [reply lib/msal-common/src/unifiedCache/generator/TokenCacheGenerator.ts]: Updated this in a following PR
